File: Analysis/GhidraScripts/CustomHeaderFieldMap.py
# @runtime PyGhidra

from ghidra.app.decompiler import DecompInterface
from ghidra.program.model.address import Address
from ghidra.program.model.lang import Register
from ghidra.program.model.scalar import Scalar


HEADER_GLOBAL = "185f2e9f8"
WINDOW_INSTRUCTIONS = 48


def targets_header_global(instruction, header_global):
    for operand in range(instruction.getNumOperands()):
        rip_relative = False
        displacement = None
        for obj in instruction.getOpObjects(operand):
            if isinstance(obj, Address) and obj == header_global:
                return True
            if isinstance(obj, Register) and obj.getBaseRegister().getName() == "RIP":
                rip_relative = True
            if isinstance(obj, Scalar):
                displacement = obj

        if rip_relative and displacement is not None:
            target = instruction.getAddress().add(instruction.getLength())
            target = target.add(displacement.getSignedValue())
            if target == header_global:
                return True

    return False


header_global = currentProgram.getAddressFactory().getDefaultAddressSpace().getAddress(HEADER_GLOBAL)
listing = currentProgram.getListing()
functions = currentProgram.getFunctionManager()
functions_by_entry = {}
function_ids = {}
ref_count = 0

print("HEADER_GLOBAL " + str(header_global))
for block in currentProgram.getMemory().getBlocks():
    if not block.isExecute():
        continue

    for source in listing.getInstructions(block, True):
        if monitor.isCancelled():
            break
        if not targets_header_global(source, header_global):
            continue

        function = functions.getFunctionContaining(source.getAddress())
        if function is None:
            continue

        entry = str(function.getEntryPoint())
        if entry not in function_ids:
            function_ids[entry] = len(function_ids) + 1
            functions_by_entry[entry] = function

        ref_count += 1
        function_id = function_ids[entry]
        print(
            "XREF {} FUNC {} {} {} AT {} {}".format(
                ref_count,
                function_id,
                entry,
                function.getName(),
                source.getAddress(),
                source,
            )
        )

        current = source
        for _ in range(WINDOW_INSTRUCTIONS):
            if current is None:
                break
            print("  {}  {}".format(current.getAddress(), current))
            current = current.getNext()

print("FUNCTION_COUNT " + str(len(function_ids)))
decompiler = DecompInterface()
decompiler.openProgram(currentProgram)
try:
    for entry, function_id in function_ids.items():
        if monitor.isCancelled():
            break

        function = functions_by_entry[entry]
        result = decompiler.decompileFunction(function, 60, monitor)
        print("DECOMP FUNC {} {} {}".format(function_id, entry, function.getName()))
        if result.decompileCompleted():
            print(result.getDecompiledFunction().getC())
        else:
            print("DECOMPILATION_FAILED " + result.getErrorMessage())
finally:
    decompiler.dispose()

File: Analysis/GhidraScripts/CustomHeaderFieldMap.java
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.lang.Register;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.scalar.Scalar;

import java.util.LinkedHashMap;
import java.util.Map;

public class CustomHeaderFieldMap extends GhidraScript {
    private static final String HEADER_GLOBAL = "185f2e9f8";
    private static final int WINDOW_INSTRUCTIONS = 32;

    private boolean targetsHeaderGlobal(Instruction instruction, Address headerGlobal) {
        for (int operand = 0; operand < instruction.getNumOperands(); operand++) {
            boolean ripRelative = false;
            Scalar displacement = null;
            for (Object object : instruction.getOpObjects(operand)) {
                if (object instanceof Address && object.equals(headerGlobal)) {
                    return true;
                }
                if (object instanceof Register
                    && ((Register) object).getBaseRegister().getName().equals("RIP")) {
                    ripRelative = true;
                }
                if (object instanceof Scalar) {
                    displacement = (Scalar) object;
                }
            }

            if (ripRelative && displacement != null) {
                Address target = instruction.getAddress().add(instruction.getLength())
                    .add(displacement.getSignedValue());
                if (target.equals(headerGlobal)) {
                    return true;
                }
            }
        }
        return false;
    }

    @Override
    public void run() throws Exception {
        Address headerGlobal = currentProgram.getAddressFactory()
            .getDefaultAddressSpace().getAddress(HEADER_GLOBAL);
        Listing listing = currentProgram.getListing();
        FunctionManager functions = currentProgram.getFunctionManager();
        Map<Address, Function> referencedFunctions = new LinkedHashMap<>();
        int referenceCount = 0;

        println("HEADER_GLOBAL " + headerGlobal);
        InstructionIterator instructions = listing.getInstructions(true);
        while (instructions.hasNext()) {
            if (monitor.isCancelled()) {
                break;
            }

            Instruction source = instructions.next();
            if (!targetsHeaderGlobal(source, headerGlobal)) {
                continue;
            }

            referenceCount++;
            Function function = functions.getFunctionContaining(source.getAddress());
            println("XREF " + referenceCount + " " + source.getAddress() + " " + source
                + " FUNC " + (function == null
                    ? "<none>"
                    : function.getName() + " @ " + function.getEntryPoint()));

            Instruction current = source;
            for (int i = 0; i < WINDOW_INSTRUCTIONS && current != null; i++) {
                println("  " + current.getAddress() + "  " + current);
                current = current.getNext();
            }

            if (function != null) {
                referencedFunctions.putIfAbsent(function.getEntryPoint(), function);
            }
        }

        println("REFERENCE_COUNT " + referenceCount);
        println("FUNCTION_COUNT " + referencedFunctions.size());

        DecompInterface decompiler = new DecompInterface();
        decompiler.openProgram(currentProgram);
        try {
            for (Map.Entry<Address, Function> entry : referencedFunctions.entrySet()) {
                if (monitor.isCancelled()) {
                    break;
                }

                Function function = entry.getValue();
                println("DECOMP " + function.getName() + " @ " + entry.getKey());
                DecompileResults result = decompiler.decompileFunction(function, 120, monitor);
                if (result.decompileCompleted()) {
                    println(result.getDecompiledFunction().getC());
                } else {
                    println("DECOMPILATION_FAILED " + result.getErrorMessage());
                }
            }
        } finally {
            decompiler.dispose();
        }
    }
}
